go-controller/pkg/ovn/default_network_controller.go

```diff
@@ -818,8 +818,14 @@ func (h *defaultNetworkControllerEventHandler) AddResource(obj interface{}, from
 		h.oc.eIPC.nodeZoneState.LockKey(node.Name)
 		h.oc.eIPC.nodeZoneState.Store(node.Name, h.oc.isLocalZoneNode(node))
 		h.oc.eIPC.nodeZoneState.UnlockKey(node.Name)
+		// add the 103 qos rule to new node's switch (we don't need to do it for node updates)
+		// (we don't need to remove this on node delete since entire node switch will get cleaned up anways)
+		err := h.oc.ensureDefaultNoRerouteQoSRules(node)
```

Comment:
that's a great catch! and totally possible this happens (I kinda had this thought to be honest and then thought I was over engineering :) ; thanks for keeping me honest here). I'll add it to update as well although I won't have a way to check if alreadydone don't try, I guess on libovsdb side it will be a no-op most of the times.

Reply:
I fixed this to also be called from update. I added a cachelookup. ptal if that works

Reply:
a slightly more efficient way would be to use a internal cache like we do for factory.NodeType, where we check if sync is needed by storing failures into a map:

```
if h.oc.isLocalZoneNode(oldNode) {
				// determine what actually changed in this update
				_, nodeSync := h.oc.addNodeFailed.Load(newNode.Name)
```

This cache lookup should be slightly faster than libovsdb cache look up, and avoids taking the nodeUpdateMutex. It might not be fast enough to matter...your call if you want to change it or not.

Reply:
yeah agreed your suggestion is neater and that's how I guess we handle it elsewhere, for now just going with the mutex lock (opened FUP issue to get back to improving this: https://github.com/ovn-org/ovn-kubernetes/issues/4361)